core/src/main/java/software/amazon/awssdk/core/util/Mimetypes.java

```diff
@@ -0,0 +1,173 @@
+
+    /** The default gzip mimetype: application/x-gzip */
+    public static final String MIMETYPE_GZIP = "application/x-gzip";
+
+    public static final String MIMETYPE_TEXT_PLAIN = "text/plain";
+
+    private static final Logger LOG = LoggerFactory.getLogger(Mimetypes.class);
+
+    private static final String MIME_TYPE_PATH = "software/amazon/awssdk/core/util/mime.types";
+
+    private static final ClassLoader CLASS_LOADER = Mimetypes.class.getClassLoader();
+
+    private static Mimetypes mimetypes = null;
+
+    /**
+     * Map that stores file extensions as keys, and the corresponding mimetype as values.
+     */
+    private Map<String, String> extensionToMimetypeMap = new HashMap<>();
+
+    private Mimetypes() {
+    }
+
+    /**
+     * Loads MIME type info from the file 'mime.types' in the classpath, if it's available.
+     */
+    public static synchronized Mimetypes getInstance() {
+        if (mimetypes != null) {
+            return mimetypes;
+        }
+
+        mimetypes = new Mimetypes();
+        InputStream stream = CLASS_LOADER.getResourceAsStream(MIME_TYPE_PATH);
+        if (stream != null) {
+            try {
+                mimetypes.loadAndReplaceMimetypes(stream);
+            } catch (IOException e) {
+                LOG.debug("Failed to load mime types from file in the classpath: mime.types", e);
+            } finally {
+                IoUtils.closeQuietly(stream, null);
+            }
+        }
+        return mimetypes;
+    }
+
+    /**
+     * Determines the mimetype of a file by looking up the file's extension in an internal listing
+     * to find the corresponding mime type. If the file has no extension, or the extension is not
+     * available in the listing contained in this class, the default mimetype
+     * <code>application/octet-stream</code> is returned.
+     * <p>
+     * A file extension is one or more characters that occur after the last period (.) in the file's name.
+     * If a file has no extension,
+     * Guesses the mimetype of file data based on the file's extension.
+     *
+     * @param file the file whose extension may match a known mimetype.
+     * @return the file's mimetype based on its extension, or a default value of
+     * <code>application/octet-stream</code> if a mime type value cannot be found.
+     */
+    public String getMimetype(File file) {
+        return getMimetype(file.getName());
+    }
+
+    /**
+     * Reads and stores the mime type setting corresponding to a file extension, by reading
+     * text from an InputStream. If a mime type setting already exists when this method is run,
+     * the mime type value is replaced with the newer one.
+     */
+    void loadAndReplaceMimetypes(InputStream is) throws IOException {
+        BufferedReader br = new BufferedReader(
+            new InputStreamReader(is, StringUtils.UTF8));
+        String line;
+
+        while ((line = br.readLine()) != null) {
```

Comment:
This might be more readable with Streams.
https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#lines-java.nio.file.Path-